discovery/premature_cache.go

```diff
@@ -0,0 +1,266 @@
+package discovery
+
+import (
+	"errors"
+	"fmt"
+	"sync"
+
+	"github.com/lightninglabs/neutrino/cache"
+	"github.com/lightninglabs/neutrino/cache/lru"
+	"github.com/lightningnetwork/lnd/lnwire"
+)
+
+// maxPrematureUpdates tracks the max amount of premature channel updates and
+// node announcements that we'll hold onto for each sender.
+const maxPrematureUpdates = 10
+
+var (
+	// errUnknownMsgType is used when unknown value type is found in the
+	// premature cache.
+	errUnknownMsgType = errors.New("unexpected msg type found in cache")
+)
+
+// messageKey defines the key to be used when saving items to premature message
+// cache.
+type messageKey interface {
+	// We use either a shortChanID for ChannelUpdate or node pubKey for
+	// NodeAnnouncement.
+	uint64 | [33]byte
+}
+
+// prematureCache defines an interface that handles reading and writing
+// premature network messages to the relaying node's cache.
+type prematureCache[T messageKey] interface {
+	// fetchMessage reads and removes the cached messages from the relaying
+	// node's cache. It takes a public key to query the node's cache and
+	// reads the value using the key.
+	fetchMessage(cacheKey [33]byte, key T) (*cachedNetworkMsg, error)
+
+	// writeMessage writes a message to the relaying node's cache. It will
+	// fetch the node's cache using the public key of the message's sender,
+	// and writes the message based on the `lnwire.Message` type.
+	writeMessage(nMsg *networkMsg) error
+}
+
+// cachedNetworkMsg is a wrapper around a network message that can be used with
+// *lru.Cache.
+type cachedNetworkMsg struct {
+	msgs []*processedNetworkMsg
+}
+
+// Size returns the "size" of an entry. We return the number of items as we
+// just want to limit the total amount of entries rather than do accurate size
+// accounting.
+//
+// NOTE: part of the cache.Value interface.
+func (c *cachedNetworkMsg) Size() (uint64, error) {
+	return uint64(len(c.msgs)), nil
+}
+
+// relayingNodeCache implements `prematureCache` interface. It is responsible
+// for managing premature network messages by each sender.
+type relayingNodeCache[T messageKey] struct {
+	sync.Mutex
+
+	// cacheMap keeps a track of caches by the relaying node's public key.
+	cacheMap map[[33]byte]*lru.Cache
```

Comment:
Ah I see now, so it's a multi-level cache, which also partitions by the target/source. This might have some benefits as well, w.r.t any adversary attempting a "cache poisoning" attack against the existing global cache.

Reply:
Exactly, this is similar to the validation deny signal fix, where we segregate each relaying nodes.